### DISCUSSION TITLE
Is it possible to accelerate cuSpatial with NVIDIA RT Cores

### DISCUSSION THREAD

**Discussion by pwrliang** [rapidsai/cuspatial#1378] [Ideas]
Hi cuSpatial community, 

Our recent work, RayJoin([paper](https://gengl.me/public/publications/ics24.pdf), [code](https://github.com/pwrliang/RayJoin)), leverages RT Cores to accelerate spatial join queries, specifically Line Segment Intersection (LSI) and Point in Polygon (PIP), by transforming spatial queries into ray tracing problems. This approach enables the queries to run on RT Cores, taking advantage of hardware acceleration. RayJoin demonstrates real-time performance, processing in milliseconds for datasets containing millions of polygons. We evaluated RayJoin against cuSpatial and other CPU/GPU spatial join solutions, noting significant performance gains. For instance, our PIP query completed in 227 milliseconds when joining two OpenStreetMap datasets with 1.6M and 303K polygons respectively on an RTX 3090, while cuSpatial took about 2120 seconds. Additionally, we implemented the LSI query on RT Cores, similar to cuSpatial’s Pairwise Linestring Intersections, but without the restriction to paired inputs, allowing it to identify arbitrary intersections between two sets of line segments.

Here are my thoughts. I believe that GIS practitioners, who are typical users of cuSpatial and often use Jupyter Notebook to interact with spatial data. A high-performance spatial join processing would enable quicker exploration of spatial data and expedite users' workflow. Furthermore, as both cuSpatial and RT Cores are NVIDIA products, integrating RT Cores-accelerated spatial queries into cuSpatial could be cool.

**Comment by jarmak-nv** [comment]
Hi @pwrliang - thanks for checking us out! 

Your approach looks and seems very interesting/fast - really love seeing RT cores getting used for more things out in the wild.

cuSpatial's perf looks pretty rough in the paper, but I'm also not too shocked. It doesn't run at top performance on the GeForce cards since we require FP64 precision. The RTX3090 has ~35.6 FP32 TFLOPS vs ~0.55 FP64 TFLOPS.

This could be an interesting way to increase our perf on but we'd need to do exploration to see how we could integrate it with the existing GeoArrow data structures and FP64 restrictions. We do have two issues ([1](https://github.com/rapidsai/cuspatial/issues/1076), [2](https://github.com/rapidsai/cuspatial/issues/1099)) open on looking into how to support FP32 but currently it's not on the horizon.

We're always happy to get feature reqs or PRs if you'd like to formalize the request!

**Reply by pwrliang** [reply]
Hi Ben, thanks for your reply. We noticed that the consumer-grade GPUs have limited FP64 ALUs. Our method works around this limitation by scaling FP64 to INT128. The intersection tests are conducted with INT128 operations, so we eliminate the expensive FP64 ops and achieve high precision. We will evaluate cuSpatial on A100 and see how the performance improves. We will also investigate the solution for integrating our RT-solution under cuSpatial's codebase.

**Comment by trxcllnt** [comment]
@pwrliang Do you have the code for the `LKEU ⊲⊳ PKEU` benchmark that OOM'd in cuSpatial? I want to check whether  https://github.com/rapidsai/cuspatial/pull/1381 fixes the issue.

I tried reconstructing the datasets from the lakes/parks [datasets link](https://datadryad.org/stash/share/aIs0nLs2TsLE_dcWO2qPHiohRKoOI3kx0WGT5BnATtA) in the RayJoin readme and running `misc/cdb2shp.py` to construct shapefiles, but I'm not sure how to transform the resulting GeoDataFrames into the point and polygon inputs required by the quadtree.

Thanks!

**Reply by pwrliang** [reply]
Running cdb2shp.py will generate the shapefile with polylines but cuSpatial's PIP requires polygons, so it may not work. Here's the LKEU_PKEU dataset I just uploaded to Google Drive: https://drive.google.com/file/d/1WEIrBG9wWDRZE-RJGnoDfQvu2eypR26Z/view?usp=sharing

Here's the code I use to run the PIP query:
```python3
import cuspatial
import cudf
import cupy
import geopandas
import pandas as pd
import numpy as np
from shapely.geometry import *
from shapely import wkt
from shapely.geometry import MultiPoint

df1 = geopandas.read_file("lakes.shp")
df1 = df1.to_crs(epsg=4326)
df1 = df1.geometry.explode(ignore_index=True)

df2 = geopandas.read_file("parks.shp")
df2 = df2.to_crs(epsg=4326)
df2 = df2.geometry.explode(ignore_index=True)

points = df2.apply(lambda x: MultiPoint(list(x.exterior.coords)))
points = points.explode(ignore_index=True)
min_x1, min_y1, max_x1, max_y1 = points.total_bounds
min_x2, min_y2, max_x2, max_y2 = df1.geometry.total_bounds
min_x = min(min_x1, min_x2)
min_y = min(min_y1, min_y2)
max_x = max(max_x1, max_x2)
max_y = max(max_y1, max_y2)

gpu_df1 = cuspatial.from_geopandas(df1)
gpu_points = cuspatial.from_geopandas(points)

max_depth = 7
max_size = 50
scale = max(max_x - min_x, max_y - min_y) // (1 << max_depth)
n_polys = len(gpu_df1)

import time
start_time = time.time()
point_indices, quadtree = cuspatial.quadtree_on_points(gpu_points,
                                                       min_x,
                                                       max_x,
                                                       min_y,
                                                       max_y,
                                                       scale,
                                                       max_depth,
                                                       max_size)
end_time = time.time()

print("Preprocessing", (end_time - start_time)*1000, "ms")

start_time = time.time()

poly_bboxes = cuspatial.polygon_bounding_boxes(
    gpu_df1
)

intersections = cuspatial.join_quadtree_and_bounding_boxes(
    quadtree,
    poly_bboxes,
    min_x,
    max_x,
    min_y,
    max_y,
    scale,
    max_depth
)
polygons_and_points = cuspatial.quadtree_point_in_polygon(
    intersections,
    quadtree,
    point_indices,
    gpu_points,
    gpu_df1
)
end_time = time.time()
print("Processing", (end_time - start_time)*1000, "ms")
```

**Reply by pwrliang** [reply]
Note. In the paper, I need to batch the polygons to avoid OOM issues (e.g., take 1000 polygons at a time until the Dataframe is exhausted), this may lead to very bad performance numbers. Hope your fix will work so the query can be processed in one batch.

**Reply by trxcllnt** [reply]
Yes, there were three bugs PR https://github.com/rapidsai/cuspatial/pull/1381 addresses:
1. A counting/overflow error due to accumulating into `uint32_t`, fixed by accumulating into `uint64_t` ([fixed here](https://github.com/rapidsai/cuspatial/blob/67c66a32c230aa3a00b19dd6f3ac3ce86ec1fa11/cpp/include/cuspatial/detail/join/quadtree_point_in_polygon.cuh#L140-L141))
2. That counting bug was masking a bug in how we were calling Thrust. `thrust::count_if` has a limit on the number of input elements it can process (`INT32_MAX`), and `thrust::copy_if`'s limit is I _think_ `UINT32_MAX`. Since we always used `thrust::copy_if` and never processed more than `UINT32_MAX` elements, we never hit this limitation. The solution is to count/copy in chunks of no larger than `INT32_MAX` ([noted here](https://github.com/rapidsai/cuspatial/blob/67c66a32c230aa3a00b19dd6f3ac3ce86ec1fa11/cpp/include/cuspatial/detail/join/quadtree_point_in_polygon.cuh#L157-L158)).
3. The routine could OOM due to attempting to allocate output buffers larger than available GPU memory. We have to over-allocate (assuming every poly will intersect with every point in the worst case), even though in most cases the intersection count will not need all that space. This over-allocation can easily OOM. The solution is to first try the over-allocation approach (because it's fastest), then fall back to computing an exact count for the output buffers ([fixed here](https://github.com/rapidsai/cuspatial/blob/67c66a32c230aa3a00b19dd6f3ac3ce86ec1fa11/cpp/include/cuspatial/detail/join/quadtree_point_in_polygon.cuh#L220-L234)).

**Reply by trxcllnt** [reply]
@pwrliang is there any reason you set `max_depth = 7, max_size = 50` in your test?

**Reply by pwrliang** [reply]
These values are just trying to make sure the code I provided works.  In the evaluation of the paper, I set `max_depth = 7` and `max_size = # of points / 4^7 / 4` as suggested by https://docs.rapids.ai/api/cuspatial/legacy/user_guide/cuspatial_api_examples/

**Reply by trxcllnt** [reply]
I can confirm PR https://github.com/rapidsai/cuspatial/pull/1381 fixes the OOM issue.

Starting from your script, I added the ability to [serialize](https://gist.github.com/trxcllnt/38e44bc86cabff23352f3541add6228f#file-rayjoin_pip-py-L15-L51) the transformed structures to Arrow tables (so rerunning is fast), [compute](https://gist.github.com/trxcllnt/38e44bc86cabff23352f3541add6228f#file-rayjoin_pip-py-L100-L106) an optimum `max_depth` and `max_size` over a user-provided threshold, and [compare](https://gist.github.com/trxcllnt/38e44bc86cabff23352f3541add6228f#file-rayjoin_pip-py-L135-L137) both single and double-precision performance.

These are the results on my 4090:

```shell
$ python rayjoin_pip.py 10
dtype: float32
  params: {'max_depth': 10, 'max_size': 18.610915184020996, 'scale': 0.0977388471364975}
  tree build (83.222ms)
  compute bboxes (34.97ms)
  num poly/quad pairs: 611328 (5.442ms)
  num intersections: 461937 (2928.46ms)
  point in polygon (2968.92ms)
  total time: 3052.172ms

dtype: float64
  params: {'max_depth': 10, 'max_size': 18.610915184020996, 'scale': 0.0977388490235464}
  tree build (73.656ms)
  compute bboxes (36.878ms)
  num poly/quad pairs: 611294 (5.507ms)
  num intersections: 464331 (17671.38ms)
  point in polygon (17713.809ms)
  total time: 17787.48ms
```

Passing a higher threshold yields a lower `max_depth`/higher `max_size`, and performance is impacted due to testing more points per quadrant against each poly. So using the largest `max_depth` and smallest `max_size` is best for performance.

**Reply by pwrliang** [reply]
Thank you for the bug fix and your benchmark scripts. I will reevaluate cuSpatial after your fix is merged into the upcoming release. This fix enables me to run PIP queries in one shot. Without this fix, I have to execute the queries in batch mode as a workaround, resulting in significantly poorer performance. Unfortunately, our paper is already camera-ready and we cannot update the performance numbers anymore, but I will post the updated performance numbers on the RayJoin GitHub page.
